### PR TITLE
Refactor workers in BuildCompilers.

### PR DESCRIPTION
**Pull request by matanlurey**
Work towards https://github.com/dart-lang/build/issues/617.

Let me know if you'd like any changes, or to make it more `bazel_worker` specific again, but this looked like the least amount of boilerplate for creating reusable workers without getting _too_ abstract. It also looks like it might be useful to integrate external tools in the future.

(Happy to add tests, even though this mostly just wraps everything for now)

### CODE REVIEW

File: build_compilers/lib/src/dev_compiler_builder.dart [hunk 1]
```diff
@@ -64,9 +64,9 @@ Future createDevCompilerModule(Module module, BuildStep buildStep,
   var jsOutputFile = scratchSpace.fileFor(module.jsId);
   var libraryRoot = '/${p.split(p.dirname(module.jsId.path)).first}';
   var sdkSummary = p.url.join(sdkDir, 'lib/_internal/ddc_sdk.sum');
-  var request = new WorkRequest();
+  var request = <String>[];
```
Comment (natebosch) [inline build_compilers/lib/src/dev_compiler_builder.dart]: [nit] rename to 'arguments' or similar

File: build_compilers/lib/src/workers.dart [hunk 2]
```diff
@@ -9,76 +9,134 @@ import 'dart:math' show min;
...
-    _analyzerWorkersAreDoneCompleter?.future ?? new Future.value(null);
-Completer<Null> _analyzerWorkersAreDoneCompleter;
+/// On windows some tools have to be executed with the `.bat` extension.
+final String _scriptExtension = Platform.isWindows ? '.bat' : '';
 
-/// Completes once the dartdevc workers have been shut down.
-Future<Null> get dartdevcWorkersAreDone =>
-    _dartdevcWorkersAreDoneCompleter?.future ?? new Future.value(null);
-Completer<Null> _dartdevcWorkersAreDoneCompleter;
+/// Returns the path to the SDK.
```
Comment (natebosch) [inline build_compilers/lib/src/workers.dart]: [nit] use a noun phrase

File: build_compilers/lib/src/workers.dart [hunk 3]
```diff
@@ -9,76 +9,134 @@ import 'dart:math' show min;
...
+    log.warning('Invalid value for $_maxWorkersEnvVar environment variable, '
+        'expected an int but got `$value`. Falling back to default value '
+        'of $_defaultMaxWorkers.');
+    return _defaultMaxWorkers;
+  },
+);
+
+/// Encapsulates running an external tool in a sandbox as a compiler.
+///
+/// TODO: Consider making this public in build_compilers(?) in the future.
```
Comment (natebosch) [inline build_compilers/lib/src/workers.dart]: I think if we were to do that we'd need to hide details like `WorkResponse`

File: build_compilers/lib/src/workers.dart [hunk 4]
```diff
@@ -9,76 +9,134 @@ import 'dart:math' show min;
...
 
-/// Resource for fetching the current [BazelWorkerDriver] for dartdevc.
-final dartdevcDriverResource = new Resource<BazelWorkerDriver>(
-    () => _dartdevcDriver, beforeExit: () async {
-  await _dartdevcDriver?.terminateWorkers();
-  _dartdevcWorkersAreDoneCompleter.complete();
-  _dartdevcWorkersAreDoneCompleter = null;
-  __dartdevcDriver = null;
-});
+class _DartDevCWorker extends BatchWorker {
```
Comment (natebosch) [inline build_compilers/lib/src/workers.dart]: Could we try a pattern without `extends`? If we don't think we'll need to `@override` other methods in the class I really like when we're precise about what behavior can possibly be changed.

```dart
typedef Future<Process> StartProcess(String workingDirectory);

class BatchWorker {
  final StartProcess _startProcess;
  ...
  BatchWorker(this._startProcess);
}

final dartDevcWorker = new BatchWorker((workingDirectory) => Process.start(
    _defaultPath, const ['--persistent_worker'],
    workingDirectory: workignDirectory));
```

File: build_compilers/lib/src/summary_builder.dart [hunk 5]
```diff
@@ -70,8 +70,8 @@ Future createUnlinkedSummary(Module module, BuildStep buildStep,
   await scratchSpace.ensureAssets(module.sources, buildStep);
 
   var summaryOutputFile = scratchSpace.fileFor(module.unlinkedSummaryId);
-  var request = new WorkRequest();
-  request.arguments.addAll([
+  var request = <String>[];
```
Comment (jakemac53) [inline build_compilers/lib/src/summary_builder.dart]: same as nate said above, lets rename this `args` or similar

File: build_compilers/lib/src/workers.dart (hunk 3 above)
Comment (jakemac53) [reply build_compilers/lib/src/workers.dart]: We could change it to a `ProcessResult`, and we could make something that wraps a work response and implements that

**Comment by matanlurey** [comment]
I'm not sure this is the right approach anymore since dart2js does not support the same batch mode as the other compilers. We can always resurrect if I'm wrong :)

**Pull request closed without merging**